### PR TITLE
Allow creation of X11 window with specified visual and screen

### PR DESCRIPTION
**Pull request by rerion**
My attempt at solving tomaka/glutin#834

### CODE REVIEW

File: src/os/unix.rs [hunk 1]
```diff
@@ -142,8 +144,22 @@ impl WindowExt for Window {
 
 /// Additional methods on `WindowBuilder` that are specific to Unix.
 pub trait WindowBuilderExt {
-
+    fn with_visual(self, visual_infos: *const XVisualInfo) -> WindowBuilder;
```
Comment (tomaka) [inline src/os/unix.rs]: Please do `with_visual<P>(self, visual_infos: *const P) -> WindowBuilder` instead, to avoid having to expose types from X11.

Comment (tomaka) [inline src/os/unix.rs]: Also please name these two methods `with_x11_visual` and `with_x11_screen`.

File: src/os/unix.rs [hunk 2]
```diff
@@ -142,8 +144,22 @@ impl WindowExt for Window {
...
+    fn with_screen(self, screen_id: i32) -> WindowBuilder;
 }
 
 impl WindowBuilderExt for WindowBuilder {
+    #[inline]
+    fn with_visual(mut self, visual_infos: *const XVisualInfo) -> WindowBuilder {
+        self.platform_specific.visual_infos = Some(
+            unsafe { ptr::read(visual_infos) }
+            );
```
Comment (tomaka) [inline src/os/unix.rs]: Nitpick: wrong indentation here

**Comment by rerion** [comment]
I've addressed those issues and also added new freestanding function for retrieving XConnection. Is that acceptable?

File: src/os/unix.rs [hunk 3]
```diff
@@ -1,17 +1,27 @@
...
 use WindowBuilder;
 use api::x11::XConnection;
+use api::x11::ffi::XVisualInfo;
 
 use wayland_client::protocol::wl_display::WlDisplay;
 use wayland_client::protocol::wl_surface::WlSurface;
 
 pub use api::x11;
 
+pub fn get_x11_xconnection() -> Option<Arc<XConnection>> {
```
Comment (tomaka) [inline src/os/unix.rs]: I don't like the fact that the `XConnection` is exposed, but the fix would be complicated so please add a `TODO: do not expose XConnection` here.

**Comment by tomaka** [comment]
Yes, just please add the small TODO and I'll merge.

**Comment by tomaka** [comment]
Also if you read this in time, please bump the version number in the Cargo.toml so that it will automatically publish the version when I merge.

File: src/os/unix.rs [hunk 4]
```diff
@@ -142,8 +153,22 @@ impl WindowExt for Window {
...
-
+    fn with_x11_visual<T>(self, visual_infos: *const T) -> WindowBuilder;
+    fn with_x11_screen(self, screen_id: i32) -> WindowBuilder;
 }
 
 impl WindowBuilderExt for WindowBuilder {
+    #[inline]
+    fn with_x11_visual<T>(mut self, visual_infos: *const T) -> WindowBuilder {
+        self.platform_specific.visual_infos = Some(
+            unsafe { ptr::read(visual_infos as *const XVisualInfo) }
```
Comment (eddyb) [inline src/os/unix.rs]: Shouldn't `with_x11_visual` be `unsafe`? I can't find any issue open about this, and I'm not sure about backwards compatibility but this seems unsound as-is.

**Pull request merged**